flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamGraph.java

```diff
@@ -333,32 +356,41 @@ public void addEdge(Integer upStreamVertexID, Integer downStreamVertexID, int ty
 				downStreamVertexID,
 				typeNumber,
 				null,
-				new ArrayList<String>());
+				new ArrayList<String>(), null);
 
 	}
 
 	private void addEdgeInternal(Integer upStreamVertexID,
 			Integer downStreamVertexID,
 			int typeNumber,
 			StreamPartitioner<?> partitioner,
-			List<String> outputNames) {
-
+			List<String> outputNames,
+			OutputTag outputTag) {
 
-		if (virtualSelectNodes.containsKey(upStreamVertexID)) {
+		if (virtualOutputNodes.containsKey(upStreamVertexID)) {
+			int virtualId = upStreamVertexID;
+			upStreamVertexID = virtualOutputNodes.get(virtualId).f0;
+			if (outputTag == null) {
+				// selections that happen downstream override earlier selections
```

Comment:
I think the comment is a leftover from copying this code from split/select. For side outputs it can't happen that you have multiple "selects" after one another. Will remove the comment. What do you think?